foreman/data_refinery_foreman/surveyor/transcriptome_index.py

```diff
@@ -0,0 +1,262 @@
+                                       species_sub_dir=self.species_sub_dir,
+                                       file_name_species=self.file_name_species,
+                                       assembly=self.assembly,
+                                       assembly_version=self.assembly_version)
+
+
+class MainEnsemblUrlBuilder(EnsemblUrlBuilder):
+    """Special logic specific to the main Ensembl division.
+
+    There is one Ensembl division which is just called Ensembl. This
+    is confusing so I refer to it as the main Ensembl division. It
+    follows the same general pattern as the rest of them for URLs, but
+    just not quite the same base URL structure. Also its REST API
+    returns JSON with similar data except with slightly different key
+    names.
+    """
+
+    def __init__(self, species: Dict):
+        self.url_root = "ensembl.org/pub/release-90"
+        self.short_division = None
+        self.species_sub_dir = species["name"]
+        self.file_name_species = species["name"].capitalize()
+        self.assembly = species["assembly"]
+        self.assembly_version = "90"
+
+        self.scientific_name = species["common_name"].upper()
+        self.taxonomy_id = species["taxon_id"]
+
+
+class EnsemblProtistsUrlBuilder(EnsemblUrlBuilder):
+    """Special logic specific to the EnsemblProtists division.
+
+    EnsemblProtists is special because the first letter of the species
+    name is always capitalized within the name of the file, instead of
+    only when there's not a collection subnested.
+    """
+
+    def __init__(self, species: Dict):
+        super().__init__(species)
+        self.file_name_species = species["species"].capitalize()
+
+
+class EnsemblFungiUrlBuilder(EnsemblProtistsUrlBuilder):
+    """The EnsemblFungi URLs work the similarly to Protists division.
+
+    EnsemblFungi is special because there is an assembly_name TIGR
+    which needs to be corrected to CADRE for some reason.
+    """
+
+    def __init__(self, species: Dict):
+        super().__init__(species)
+        if self.assembly == "TIGR":
+            self.assembly = "CADRE"
+
+
+def ensembl_url_builder_factory(species: Dict) -> EnsemblUrlBuilder:
+    """Returns instance of EnsemblUrlBuilder or one of its subclasses.
+
+    The class of the returned object is based on the species' division.
+    """
+    if species["division"] == "EnsemblProtists":
+        return EnsemblProtistsUrlBuilder(species)
+    elif species["division"] == "EnsemblFungi":
+        return EnsemblFungiUrlBuilder(species)
+    elif species["division"] == "Ensembl":
+        return MainEnsemblUrlBuilder(species)
+    else:
+        return EnsemblUrlBuilder(species)
+
+
+class TranscriptomeIndexSurveyor(ExternalSourceSurveyor):
+    def source_type(self):
+        return Downloaders.TRANSCRIPTOME_INDEX.value
+
+    def determine_pipeline(self, batch: Batch, key_values: Dict = {}) -> ProcessorPipeline:
+        return ProcessorPipeline.TRANSCRIPTOME_INDEX
```

Comment:
The return value actually is not a string. `ProcessorPipeline.TRANSCRIPTOME_INDEX.value` is a string, but it turns out that Enums in python are somewhat confusing.

```python
>>> type(ProcessorPipeline.TRANSCRIPTOME_INDEX)
<enum 'ProcessorPipeline'>
```

```python
>>> inspect.getmro(type(ProcessorPipeline.TRANSCRIPTOME_INDEX))
(<enum 'ProcessorPipeline'>, <enum 'PipelineEnums'>, <enum 'Enum'>, <class 'object'>)
```

Reply:
It kinda makes sense though because how else could you express the intent "this function returns one of the valid values for the ProcessorPipeline Enum"?